### PR TITLE
T-WasmVm-3: Gatekeeper

### PR DESCRIPTION
**Pull request by Rhaki**
Added tests for Gatekeeper

### CODE REVIEW

File: crates/vm/wasm/src/gatekeeper.rs [hunk 1]
```diff
@@ -749,3 +749,73 @@ impl FunctionMiddleware for FunctionGatekeeper {
         }
     }
 }
+
```
Comment (larry0x) [inline crates/vm/wasm/src/gatekeeper.rs:752]: Can you always add this "separator" comment before the `tests` module? I have this for all other tests:

```rust
// ----------------------------------- test ------------------------------------
```

File: crates/vm/wasm/src/gatekeeper.rs [hunk 2]
```diff
@@ -749,3 +749,73 @@ impl FunctionMiddleware for FunctionGatekeeper {
...
+        let deterministic = Arc::new(Gatekeeper::default());
+        let mut compiler = Singlepass::new();
+        compiler.push_middleware(deterministic);
+        let store = Store::new(compiler);
+        let result = Module::new(&store, wasm);
+        assert!(result
+            .unwrap_err()
+            .to_string()
+            .contains("Bulk memory operation"));
+    }
```
Comment (larry0x) [inline crates/vm/wasm/src/gatekeeper.rs:820]: Can you use `#[test_case]` for these?

**Comment by Rhaki** [comment]
@larry0x suggestions applied

File: crates/vm/wasm/src/gatekeeper.rs [hunk 3]
```diff
@@ -763,59 +763,60 @@ mod tests {
...
-                    i32.add
-                ))
-            "#,
+          (module
+            (func (export "sum") (param i32 i32) (result i32)
+              local.get 0
+              local.get 1
+              i32.add
+            ))
+        "#,
```
Comment (larry0x) [inline crates/vm/wasm/src/gatekeeper.rs:772]: WAT uses two space indentation

File: crates/vm/wasm/src/gatekeeper.rs [hunk 4]
```diff
@@ -763,59 +763,60 @@ mod tests {
...
         |result| {
             assert!(result.is_ok());
         };
-        "valid_wasm_instance_sanity"
+        "valid wasm instance sanity"
```
Comment (larry0x) [inline crates/vm/wasm/src/gatekeeper.rs:776]: Test name should be a human readable string (see the example usage in test-case: https://docs.rs/test-case/latest/test_case/#example-usage), so no need for the underscores

File: crates/vm/wasm/src/gatekeeper.rs [hunk 5]
```diff
@@ -763,59 +763,60 @@ mod tests {
...
     {
-        let deterministic = Arc::new(Gatekeeper::default());
         let mut compiler = Singlepass::new();
-        compiler.push_middleware(deterministic);
+        compiler.push_middleware(Arc::new(Gatekeeper::default()));
+
         let store = Store::new(compiler);
         let result = Module::new(&store, wat);
+
         callback(result);
```
Comment (larry0x) [inline crates/vm/wasm/src/gatekeeper.rs:820]: Adding some empty lines for better readability

**Pull request merged**